### PR TITLE
Improve client tests

### LINKED ISSUE #28: Test attack scenarios
The client tests need improving to cover the attack scenarios discussed in [section 1.5.2 of the TUF spec](https://github.com/theupdateframework/tuf/blob/v0.9.9/docs/tuf-spec.txt#L121-L155) (there are some good examples in the Python test suite).

### PR DESCRIPTION
**Pull request by lmars**
This covers all TUF attack scenarios except "Slow Retrieval Attacks" (see #48).

Closes #28.

### CODE REVIEW

File: repo.go [hunk 1]
```diff
@@ -301,6 +301,40 @@ func (r *Repo) Sign(name string) error {
 	return r.local.SetMeta(name, b)
 }
 
+// getKeys returns signing keys from local storage.
+//
+// Only keys contained in the keys db are returned (i.e. local keys which have
+// been revoked are ommitted), except for the root role in which case all local
```
Comment (titanous) [inline repo.go]: s/ommitted/omitted/

File: client/client.go [hunk 2]
```diff
@@ -343,29 +352,24 @@ func (c *Client) downloadMeta(name string, m data.FileMeta) ([]byte, error) {
...
-// verifyRoot verifies root metadata.
-func (c *Client) verifyRoot(b json.RawMessage) (err error) {
-	s := &data.Signed{}
-	if err = json.Unmarshal(b, s); err != nil {
-		return
+// decodeRoot decodes and verifies root metadata.
+func (c *Client) decodeRoot(b json.RawMessage) error {
+	root := &data.Root{}
+	if err := signed.Unmarshal(b, root, "root", c.rootVer, c.db); err != nil {
+		return ErrDownloadFailed{"root.json", err}
```
Comment (titanous) [inline client/client.go]: Is `ErrDownloadFailed` correct here? Perhaps it should have a comment?

File: client/client.go [hunk 3]
```diff
@@ -373,11 +377,8 @@ func (c *Client) decodeSnapshot(b json.RawMessage) (data.FileMeta, data.FileMeta
 // returns updated targets.
 func (c *Client) decodeTargets(b json.RawMessage) (data.Files, error) {
 	targets := &data.Targets{}
-	if err := signed.Unmarshal(b, targets, "targets", c.localTargetsVer, c.db); err != nil {
-		if err == signed.ErrExpired {
-			err = ErrExpiredMeta{"targets.json"}
-		}
-		return nil, err
+	if err := signed.Unmarshal(b, targets, "targets", c.targetsVer, c.db); err != nil {
+		return nil, ErrDownloadFailed{"targets.json", err}
```
Comment (titanous) [inline client/client.go]: Ditto on the error.

File: client/client_test.go [hunk 4]
```diff
@@ -397,15 +543,90 @@ func (s *ClientSuite) TestUpdateRemoteExpired(c *C) {
...
+	s.addRemoteTarget(c, "bar.txt")
+	newTargets, ok := s.remote["targets.json"]
+	if !ok {
+		c.Fatal("missing remote targets.json")
+	}
+	s.remote["targets.json"] = oldTargets
+
+	// check update returns ErrWrongSize for targets.json
+	_, err := client.Update()
+	c.Assert(err, DeepEquals, ErrWrongSize{"targets.json", oldTargets.size, newTargets.size})
```
Comment (titanous) [inline client/client_test.go]: Can we also do a test where the size is the same?

**Comment by titanous** [comment]
I noticed one code path in the coverage profile that looks like it deserves a test:

```go
// if the latest root.json has been downloaded and it
// is still expired, halt the update with an error
if latestRoot {
        return nil, ErrDownloadFailed{"root.json", err}
}
```

It's probably worth going through the coverage profile and making an issue with a checklist of other code paths that look like they should be tested.

**Comment by lmars** [comment]
@titanous comments addressed.

The code path you found to be untested is actually unreachable with the current code (see [this comment](https://github.com/flynn/go-tuf/blob/improve-client-tests/client/client.go#L124-L126)). I've decided to leave it as a returned error (rather than say `panic("unreached")`) as code changes in the future could make it reachable.

**Comment by titanous** [comment]
LGTM, nice work!

**Pull request merged**